cloudflare/resource_cloudflare_authenticated_origin_pulls_certificate.go

```diff
@@ -0,0 +1,191 @@
+		},
+	}
+}
+
+func resourceCloudflareAuthenticatedOriginPullsCertificateCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*cloudflare.API)
+	zoneID := d.Get("zone_id").(string)
+	log.Printf("[DEBUG] zone ID: %s", zoneID)
+
+	switch aopType, ok := d.GetOk("type"); ok {
+	case aopType == "per-zone":
+		perZoneAOPCert := cloudflare.PerZoneAuthenticatedOriginPullsCertificateParams{
+			Certificate: d.Get("certificate").(string),
+			PrivateKey:  d.Get("private_key").(string),
+		}
+		record, err := client.UploadPerZoneAuthenticatedOriginPullsCertificate(zoneID, perZoneAOPCert)
+		if err != nil {
+			return fmt.Errorf("Error uploading Per-Zone AOP certificate on zone %q: %s", zoneID, err)
+		}
+		d.SetId(record.ID)
+
+		return resource.Retry(d.Timeout(schema.TimeoutCreate), func() *resource.RetryError {
+			resp, err := client.GetPerZoneAuthenticatedOriginPullsCertificateDetails(zoneID, record.ID)
+			if err != nil {
+				return resource.NonRetryableError(fmt.Errorf("Error reading Per Zone AOP certificate details: %s", err))
+			}
+
+			if resp.Status != "active" {
+				return resource.RetryableError(fmt.Errorf("Expected Per Zone AOP certificate to be active but was in state %s", resp.Status))
+			}
+
+			return resource.NonRetryableError(resourceCloudflareAuthenticatedOriginPullsCertificateRead(d, meta))
+		})
+	case aopType == "per-hostname":
+		perHostnameAOPCert := cloudflare.PerHostnameAuthenticatedOriginPullsCertificateParams{
+			Certificate: d.Get("certificate").(string),
+			PrivateKey:  d.Get("private_key").(string),
+		}
+		record, err := client.UploadPerHostnameAuthenticatedOriginPullsCertificate(zoneID, perHostnameAOPCert)
+		if err != nil {
+			return fmt.Errorf("Error uploading Per-Hostname AOP certificate on zone %q: %s", zoneID, err)
+		}
+		d.SetId(record.ID)
+
+		return resource.Retry(d.Timeout(schema.TimeoutCreate), func() *resource.RetryError {
+			resp, err := client.GetPerHostnameAuthenticatedOriginPullsCertificate(zoneID, record.ID)
+			if err != nil {
+				return resource.NonRetryableError(fmt.Errorf("Error reading Per Hostname AOP certificate details: %s", err))
+			}
+
+			if resp.Status != "active" {
+				return resource.RetryableError(fmt.Errorf("Expected Per Hostname AOP certificate to be active but was in state %s", resp.Status))
+			}
+
+			return resource.NonRetryableError(resourceCloudflareAuthenticatedOriginPullsCertificateRead(d, meta))
+		})
+	}
+	return nil
+}
+
+func resourceCloudflareAuthenticatedOriginPullsCertificateRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*cloudflare.API)
+	zoneID := d.Get("zone_id").(string)
+	certID := d.Id()
+	log.Printf("[DEBUG] zone ID: %s", zoneID)
+
+	switch aopType, ok := d.GetOk("type"); ok {
+	case aopType == "per-zone":
+		record, err := client.GetPerZoneAuthenticatedOriginPullsCertificateDetails(zoneID, certID)
+		if err != nil {
+			log.Printf("[WARN] Removing certificate from state because it was not found in API.")
```

Comment:
Is this a safe assumption to make for all exceptions we may encounter? I.e. what happens if we encounter a service outage? Other examples of this have been restricted to 404s where the resource is no loner expected.

Reply:
Ah good point. We can't always assume that a failed Read means that the certificate doesn't exist. Let me poke around for examples in other resources to see what others do.